File: Source/app/src/main/java/com/amgregoire/mangafeed/Common/WebSources/FunManga2.java
package com.amgregoire.mangafeed.Common.WebSources;


import com.amgregoire.mangafeed.Common.MangaEnums;
import com.amgregoire.mangafeed.Common.RequestWrapper;
import com.amgregoire.mangafeed.Common.WebSources.Base.SourceManga;
import com.amgregoire.mangafeed.MangaFeed;
import com.amgregoire.mangafeed.Models.DbChapter;
import com.amgregoire.mangafeed.Models.DbManga;
import com.amgregoire.mangafeed.Utils.MangaDB;
import com.amgregoire.mangafeed.Utils.MangaLogger;
import com.amgregoire.mangafeed.v2.service.Logger;
import com.google.gson.Gson;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class FunManga2 extends SourceManga
{
    public final static String TAG = FunManga.class.getSimpleName();
    public final static String URL = "https://www.funmanga.com/";

    private final String SourceKey = "FunManga";
    private final String mBaseUrl = "https://www.funmanga.com/";
    private final String mUpdatesUrl = "https://funmanga.com/latest-chapters/";
    private final String mCatalogUrl = "https://funmanga.com/manga-list/";
    private final String mGenres[] =
            {
                    "Joy",
                    "Action",
                    "Adult",
                    "Adventure",
                    "Comedy",
                    "Doujinshi",
                    "Drama",
                    "Ecchi",
                    "Fantasy",
                    "Gender Bender",
                    "Harem",
                    "Historical",
                    "Horror",
                    "Josei",
                    "Lolicon",
                    "Manga",
                    "Manhua",
                    "Manhwa",
                    "Martial Arts",
                    "Mature",
                    "Mecha",
                    "Mystery",
                    "One shot",
                    "Psychological",
                    "Romance",
                    "School Life",
                    "Sci fi",
                    "Seinen",
                    "Shotacon",
                    "Shoujo",
                    "Shoujo Ai",
                    "Shounen",
                    "Shounen Ai",
                    "Slice of Life",
                    "Smut",
                    "Sports",
                    "Supernatural",
                    "Tragedy",
                    "Yaoi",
                    "Yuri"
            };

    @Override
    public boolean requiresCloudFlare()
    {
        return true;
    }

    @Override
    public String getSourceName()
    {
        return SourceKey;
    }

    @Override
    public String getBaseUrl()
    {
        return mBaseUrl;
    }

    @Override
    public MangaEnums.SourceType getSourceType()
    {
        return MangaEnums.SourceType.MANGA;
    }

    @Override
    public String getRecentUpdatesUrl()
    {
        return mUpdatesUrl;
    }

    @Override
    public String[] getGenres()
    {
        return mGenres;
    }

    @Override
    public List<DbManga> parseResponseToRecentList(final String responseBody)
    {
        List<DbManga> lDbMangaList = new ArrayList<>();

        try
        {
            Document lParsedDocument = Jsoup.parse(responseBody);
            Elements lMangaElements = lParsedDocument.select("div.manga_updates").select("dl");

            for (Element iWholeElement : lMangaElements)
            {
                Document lParseSections = Jsoup.parse(iWholeElement.toString());
                Elements lUsefulElements = lParseSections.select("dt");
                for (Element iUsefulElement : lUsefulElements)
                {
                    String lMangaTitle = iUsefulElement.select("a").attr("title");
                    String lMangaUrl = iUsefulElement.select("a").attr("href");

                    if (lMangaUrl.charAt(lMangaUrl.length() - 1) != '/')
                    {
                        lMangaUrl += "/"; //add ending slash to url if missing
                    }

                    lMangaUrl = lMangaUrl.replaceFirst(DbManga.Companion.getLinkRegex(), "{" + SourceKey + "}");
                    DbManga lDbManga = MangaDB.getInstance().getManga(lMangaUrl);
                    if (lDbManga != null)
                    {
                        lDbMangaList.add(lDbManga);
                    }
                    else
                    {
                        lDbManga = MangaDB.getInstance().putManga(new DbManga(lMangaTitle, lMangaUrl, SourceKey));
                        lDbMangaList.add(lDbManga);

                        updateMangaObservable(new RequestWrapper(lDbManga))
                                .subscribe
                                        (
                                                manga -> MangaLogger.logInfo(TAG, "Finished updating " + manga.getTitle()),
                                                throwable -> MangaLogger.logError(TAG, "Problem updating: " + throwable
                                                        .getMessage())
                                        );
                    }
                }
            }
        }
        catch (Exception aException)
        {
//            MangaLogger.logError(TAG, responseBody);
            MangaLogger.logError(TAG, " Failed to parse recent updates: " + aException.getMessage());
        }

        MangaLogger.logInfo(TAG, "Finished parsing recent updates");

        if (lDbMangaList.size() == 0)
        {
            return null;
        }
        return lDbMangaList;
    }

    @Override
    public DbManga parseResponseToManga(final RequestWrapper request, final String responseBody)
    {
        Document lHtml = Jsoup.parse(responseBody);

        try
        {
            Element lImageElement = lHtml.body().select("img.img-responsive.mobile-img").first();
            Element lDescriptionElement = lHtml.body()
                                               .select("div.note.note-default.margin-top-15")
                                               .first();
            Elements lInfo = lHtml.body().select("dl.dl-horizontal").select("dd");

            String lImage = "";
            if (lImageElement != null && lImageElement.hasAttr("src")) lImage = lImageElement.attr("src");


            String lDescription = "";
            if (lDescriptionElement != null) lDescription = lDescriptionElement.text();
            String lAlternate = null;
            String lAuthor = null;
            String lArtist = null;
            String lGenres = null;
            String lStatus = null;


            if (lInfo != null)
            {
                for (int i = 0; i < lInfo.size(); i++)
                {
                    if (i == 0)
                    {
                        lAlternate = lInfo.get(i).text();
                    }
                    else if (i == 5)
                    {
                        lAuthor = lInfo.get(i).text();
                    }
                    else if (i == 4)
                    {
                        lArtist = lInfo.get(i).text();
                    }
                    else if (i == 2)
                    {
                        lGenres = lInfo.get(i).text();
                    }
                    else if (i == 1)
                    {
                        lStatus = lInfo.get(i).text();
                    }
                }
            }

            DbManga lDbManga = MangaDB.getInstance().getManga(request.getManga().getLink());
            lDbManga.setAlternate(lAlternate);
            lDbManga.setImage(lImage);
            lDbManga.setDescription(lDescription);
            lDbManga.setArtist(lArtist);
            lDbManga.setAuthor(lAuthor);
            lDbManga.setGenres(lGenres);
            lDbManga.setStatus(lStatus);
            lDbManga.setSource(SourceKey);
            lDbManga.setLink(request.getManga().getLink());

            MangaDB.getInstance().putManga(lDbManga);

            MangaLogger.logInfo(TAG, "Finished creating/updating manga (" + lDbManga.getTitle() + ")");
        }
        catch (Exception aException)
        {
            MangaLogger.logError(TAG, request.getManga().getLink(), aException.getMessage());
            Logger.INSTANCE.error(aException, "");
        }

        return MangaDB.getInstance().getManga(request.getManga().getLink());
    }

    @Override
    public List<DbChapter> parseResponseToChapters(RequestWrapper request, String responseBody)
    {
        Document lParsedDocument = Jsoup.parse(responseBody);
        List<DbChapter> lDbChapterList = resolveChaptersFromParsedDocument(lParsedDocument, request);

        return lDbChapterList;
    }

    @Override
    public List<String> parseResponseToPageUrls(final RequestWrapper requestWrapper, final String responseBody)
    {
        List<String> lImages = new ArrayList<>();

        Document lDoc = Jsoup.parse(responseBody);
        Elements lNav = lDoc.select("h5.widget-heading").select("select").select("option");

        int lPages = lNav.size();

        for (int i = 1; i < lPages; i++)
        {
            String lLink = lNav.get(i).attr("value");
            lImages.add(lLink);
        }

        return lImages;
    }

    @Override
    public String parseResponseToImageUrls(final String responseBody, final String responseUrl)
    {
        Document lParsedDocument = Jsoup.parse(responseBody);
        String lLink = lParsedDocument.select("img.img-responsive").attr("src");

        return lLink;
    }

    @Override
    public void updateLocalCatalog()
    {
        List<Observable> pages = new ArrayList<>();
        char lEndPoint = ' ';

        while (lEndPoint != 'z')
        {
            pages.add(updateCatalogObservable(mCatalogUrl + lEndPoint));

            if (lEndPoint == ' ')
            {
                lEndPoint = 'a';
            }
            else
            {
                lEndPoint++;
            }
        }

        Single temp = Observable.fromIterable(pages)
                                .subscribeOn(Schedulers.computation())
                                .flatMap(obs -> obs.subscribeOn(Schedulers.computation()))
                                .flatMap(response -> test((String) response))
                                .toList();

        temp.subscribe(o -> {
            String x = new Gson().toJson(o);
            Logger.INSTANCE.error(x, "");
        }, new Consumer<Throwable>()
        {
            @Override
            public void accept(Throwable throwable) throws Exception
            {
                Logger.INSTANCE.error(throwable, "oh no");
            }
        });
    }


    public List<DbManga> test(String response)
    {
        Logger.INSTANCE.error("Starting local catalog", "");
        ArrayList<DbManga> result = new ArrayList<>();

        String responseBody = (String) response;
        MangaDB lDatabase = MangaDB.getInstance();
        Document lParsedDocument = Jsoup.parse(responseBody);
        Elements lLinks = lParsedDocument.select("ul.manga-list.circle-list").select("a");

        for (Element link : lLinks)
        {
            String name = link.text();
            String url = link.attr("href") + "/";

            DbManga newDbManga = new DbManga(name, url, SourceKey);
            if (!lDatabase.containsManga(newDbManga))
            {
                newDbManga = lDatabase.putManga(newDbManga);
                Logger.INSTANCE.info("Added new item -> " + newDbManga.getTitle());
                result.add(newDbManga);
            }
        }

        return result;
    }

    //    @Override
    public void updateLocalCatalog2()
    {
        char lEndPoint = ' ';

        while (lEndPoint != 'z')
        {

            updateCatalogObservable(mCatalogUrl + lEndPoint).subscribe(response ->
                    {
                        Logger.INSTANCE.error("Starting local catalog", "");
                        String responseBody = (String) response;
                        MangaDB lDatabase = MangaDB.getInstance();
                        Document lParsedDocument = Jsoup.parse(responseBody);
                        Elements lLinks = lParsedDocument.select("ul.manga-list.circle-list").select("a");

                        for (Element link : lLinks)
                        {
                            String name = link.text();
                            String url = link.attr("href") + "/";

                            DbManga newDbManga = new DbManga(name, url, SourceKey);
                            if (!lDatabase.containsManga(newDbManga))
                            {
                                newDbManga = lDatabase.putManga(newDbManga);
                                Logger.INSTANCE.info("Added new item -> " + newDbManga.getTitle());
                                // update new entry info

                                MangaFeed.Companion.getApp()
                                                   .getSourceByTag(TAG)
                                                   .updateMangaObservable(new RequestWrapper(newDbManga))
                                                   .subscribe(
                                                           manga -> Logger.INSTANCE.info("Finished updating (" + TAG + ") " + manga.getTitle()),
                                                           throwable -> MangaLogger.logError(TAG, "Problem updating: " + throwable
                                                                   .getMessage())
                                                   );
                            }
                        }
                    },
                    throwable -> MangaLogger.logError(TAG, throwable.toString()));

            if (lEndPoint == ' ')
            {
                lEndPoint = 'a';
            }
            else
            {
                lEndPoint++;
            }
        }


    }

    /***
     * This helper function resolves chapters from the specified document and returns a list of chapters.
     * Parent - parseResponseToChapters();
     *
     * @param parsedDocument
     * @return
     */
    private List<DbChapter> resolveChaptersFromParsedDocument(final Document parsedDocument, final RequestWrapper request)
    {
        List<DbChapter> lDbChapterList = new ArrayList<>();
        Elements lChapterElements = parsedDocument.select("ul.chapter-list").select("li");
        int lNumChapters = lChapterElements.size();

        String lChapterUrl, lChapterTitle, lChapterDate;

        for (Element iChapterElement : lChapterElements)
        {
            lChapterUrl = iChapterElement.select("a").attr("href");
            lChapterTitle = iChapterElement.select("span").first().text();
            lChapterDate = iChapterElement.select("span").get(1).text();

            lDbChapterList.add(new DbChapter(lChapterUrl, request.getManga().getTitle(), lChapterTitle, lChapterDate, lNumChapters, request.getManga()
                                                                                                                                           .getLink(), SourceKey));

            lNumChapters--;
        }

        return lDbChapterList;
    }
}

File: Source/app/src/main/java/com/amgregoire/mangafeed/Common/WebSources/FunManga2.kt
package com.amgregoire.mangafeed.Common.WebSources


import com.amgregoire.mangafeed.Common.MangaEnums
import com.amgregoire.mangafeed.Common.RequestWrapper
import com.amgregoire.mangafeed.Common.SyncStatusObject
import com.amgregoire.mangafeed.Common.WebSources.Base.SourceManga
import com.amgregoire.mangafeed.MangaFeed
import com.amgregoire.mangafeed.Models.DbChapter
import com.amgregoire.mangafeed.Models.DbManga
import com.amgregoire.mangafeed.Utils.MangaDB
import com.amgregoire.mangafeed.Utils.MangaLogger
import com.amgregoire.mangafeed.v2.service.Logger
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.*

class FunManga : SourceManga()
{

    private val mBaseUrl = "https://www.funmanga.com"
    private val mUpdatesUrl = "https://funmanga.com/latest-chapters"
    private val mCatalogUrl = "https://funmanga.com/manga-list/"
    private val mGenres = arrayOf("Joy", "Action", "Adult", "Adventure", "Comedy", "Doujinshi", "Drama", "Ecchi", "Fantasy", "Gender Bender", "Harem", "Historical", "Horror", "Josei", "Lolicon", "Manga", "Manhua", "Manhwa", "Martial Arts", "Mature", "Mecha", "Mystery", "One shot", "Psychological", "Romance", "School Life", "Sci fi", "Seinen", "Shotacon", "Shoujo", "Shoujo Ai", "Shounen", "Shounen Ai", "Slice of Life", "Smut", "Sports", "Supernatural", "Tragedy", "Yaoi", "Yuri")

    override fun requiresCloudFlare(): Boolean
    {
        return true
    }

    override fun getSourceName(): String
    {
        return SourceKey
    }

    override fun getBaseUrl(): String
    {
        return mBaseUrl
    }

    override fun getSourceType(): MangaEnums.SourceType
    {
        return MangaEnums.SourceType.MANGA
    }

    override fun getRecentUpdatesUrl(): String
    {
        return mUpdatesUrl
    }

    override fun getGenres(): Array<String>
    {
        return mGenres
    }

    override fun parseResponseToRecentList(responseBody: String): List<DbManga>?
    {
        val lMangaList = ArrayList<DbManga>()

        try
        {
            val lParsedDocument = Jsoup.parse(responseBody)
            val lMangaElements = lParsedDocument.select("div.manga_updates").select("dl")

            for (iWholeElement in lMangaElements)
            {
                val lParseSections = Jsoup.parse(iWholeElement.toString())
                val lUsefulElements = lParseSections.select("dt")
                for (iUsefulElement in lUsefulElements)
                {
                    val lMangaTitle = iUsefulElement.select("a").attr("title")
                    var lMangaUrl = iUsefulElement.select("a").attr("href")

                    if (lMangaUrl[lMangaUrl.length - 1] != '/')
                    {
                        lMangaUrl += "/" //add ending slash to url if missing
                    }

                    lMangaUrl = lMangaUrl.replaceFirst(DbManga.LinkRegex.toRegex(), "{$SourceKey}")
                    var lDbManga: DbManga? = MangaDB.getInstance().getManga(lMangaUrl)
                    if (lDbManga != null)
                    {
                        lMangaList.add(lDbManga)
                    }
                    else
                    {
                        lDbManga = MangaDB.getInstance().putManga(DbManga(lMangaTitle, lMangaUrl, SourceKey))
                        lMangaList.add(lDbManga)

                        updateMangaObservable(RequestWrapper(lDbManga))
                                .subscribe(
                                        { manga -> MangaLogger.logInfo(TAG, "Finished updating " + manga.title) },
                                        { throwable ->
                                            MangaLogger.logError(TAG, "Problem updating: " + throwable
                                                    .message)
                                        }
                                )
                    }
                }
            }
        }
        catch (aException: Exception)
        {
            //            MangaLogger.logError(TAG, responseBody);
            MangaLogger.logError(TAG, " Failed to parse recent updates: " + aException.message)
        }

        MangaLogger.logInfo(TAG, "Finished parsing recent updates")

        return if (lMangaList.size == 0)
        {
            null
        }
        else lMangaList
    }

    override fun parseResponseToManga(request: RequestWrapper, responseBody: String): DbManga
    {
        val lHtml = Jsoup.parse(responseBody)

        try
        {
            val lImageElement = lHtml.body().select("img.img-responsive.mobile-img").first()
            val lDescriptionElement = lHtml.body()
                    .select("div.note.note-default.margin-top-15")
                    .first()
            val lInfo = lHtml.body().select("dl.dl-horizontal").select("dd")

            var lImage = ""
            if (lImageElement != null && lImageElement.hasAttr("src")) lImage = lImageElement.attr("src")


            var lDescription = ""
            if (lDescriptionElement != null) lDescription = lDescriptionElement.text()
            var lAlternate: String? = null
            var lAuthor: String? = null
            var lArtist: String? = null
            var lGenres: String? = null
            var lStatus: String? = null


            if (lInfo != null)
            {
                for (i in lInfo.indices)
                {
                    when (i)
                    {
                        0 -> lAlternate = lInfo[i].text()
                        5 -> lAuthor = lInfo[i].text()
                        4 -> lArtist = lInfo[i].text()
                        2 -> lGenres = lInfo[i].text()
                        1 -> lStatus = lInfo[i].text()
                    }
                }
            }

            val lManga = MangaDB.getInstance().getManga(request.manga.link)
            lManga.alternate = lAlternate
            lManga.image = lImage
            lManga.description = lDescription
            lManga.artist = lArtist
            lManga.author = lAuthor
            lManga.genres = lGenres
            lManga.status = lStatus
            lManga.source = SourceKey
            lManga.link = request.manga.link

            MangaDB.getInstance().putManga(lManga)

            MangaLogger.logInfo(TAG, "Finished creating/updating manga (" + lManga.title + ")")
        }
        catch (aException: Exception)
        {
            MangaLogger.logError(TAG, request.manga.link, aException.message)
            Logger.error(aException, "")
        }

        return MangaDB.getInstance().getManga(request.manga.link)
    }

    override fun parseResponseToChapters(request: RequestWrapper, responseBody: String): List<DbChapter>
    {
        val lParsedDocument = Jsoup.parse(responseBody)

        return resolveChaptersFromParsedDocument(lParsedDocument, request)
    }

    override fun parseResponseToPageUrls(requestWrapper: RequestWrapper, responseBody: String): List<String>
    {
        val lImages = ArrayList<String>()

        val lDoc = Jsoup.parse(responseBody)
        val lNav = lDoc.select("h5.widget-heading").select("select").select("option")

        val lPages = lNav.size

        for (i in 1 until lPages)
        {
            val lLink = lNav[i].attr("value")
            lImages.add(lLink)
        }

        return lImages
    }

    override fun parseResponseToImageUrls(responseBody: String, responseUrl: String): String
    {
        val lParsedDocument = Jsoup.parse(responseBody)

        return lParsedDocument.select("img.img-responsive").attr("src")
    }

    /***
     * New way to update catalog, provides feedback to ui to gage activity.
     * Currently only added for FunManga
     */
    override fun updateLocalCatalogV2(): Observable<SyncStatusObject>
    {
        val endPoints = listOf(' ', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z')
        val pages = endPoints.map { char -> updateCatalogObservable(mCatalogUrl + char) }
        var counter = 0
        return Observable.fromIterable(pages)
                .subscribeOn(Schedulers.computation())
                .flatMap { obs -> obs.subscribeOn(Schedulers.computation()) }
                .map { response ->
                    counter++
                    Pair(counter, response)
                }
                .map { (count, response) -> Pair(count, convertCatalogPageToMangaList(response as String)) }
                .flatMap { (count, list) -> Observable.just(SyncStatusObject(count, endPoints.size, list)) }
    }


    private fun convertCatalogPageToMangaList(response: String): List<DbManga>
    {
        Logger.error("Starting local catalog", "")
        val result = ArrayList<DbManga>()

        val lDatabase = MangaDB.getInstance()
        val lParsedDocument = Jsoup.parse(response)
        val lLinks = lParsedDocument.select("ul.manga-list.circle-list").select("a")

        for (link in lLinks)
        {
            val name = link.text()
            val url = link.attr("href") + "/"

            //            var newManga = Manga(name, url, SourceKey)
            //            if (!lDatabase.containsManga(newManga))
            //            {
            //                newManga = lDatabase.putManga(newManga)
            //                Logger.info("Added new item -> " + newManga.title)
            //                result.add(newManga)
            //            }
            result.add(DbManga(name, url, SourceKey))
        }

        val links = result.map { m -> m.link }.chunked(200)
        val existing = arrayListOf<DbManga>()

        for (list in links)
        {
            existing.addAll(lDatabase.getExistingManga(list, sourceName))
        }

        result.removeAll(existing)

        result.map { old ->
            Logger.info("Added new item -> " + old.title)
            lDatabase.putManga(old)
        }

        return result.toList()
    }

    override fun updateLocalCatalog()
    {
        var lEndPoint = ' '

        while (lEndPoint != 'z')
        {

            updateCatalogObservable(mCatalogUrl + lEndPoint).subscribe({ response ->
                Logger.error("Starting local catalog", "")
                val responseBody = response as String
                val lDatabase = MangaDB.getInstance()
                val lParsedDocument = Jsoup.parse(responseBody)
                val lLinks = lParsedDocument.select("ul.manga-list.circle-list").select("a")

                for (link in lLinks)
                {
                    val name = link.text()
                    val url = link.attr("href") + "/"

                    var newManga = DbManga(name, url, SourceKey)
                    if (!lDatabase.containsManga(newManga))
                    {
                        newManga = lDatabase.putManga(newManga)
                        Logger.info("Added new item -> " + newManga.title)
                        // update new entry info

                        MangaFeed.app
                                .getSourceByTag(TAG)
                                .updateMangaObservable(RequestWrapper(newManga))
                                .subscribe(
                                        { manga -> Logger.info("Finished updating (" + TAG + ") " + manga.title) },
                                        { throwable ->
                                            MangaLogger.logError(TAG, "Problem updating: " + throwable
                                                    .message)
                                        }
                                )
                    }
                }
            },
                    { throwable -> MangaLogger.logError(TAG, throwable.toString()) })

            if (lEndPoint == ' ')
            {
                lEndPoint = 'a'
            }
            else
            {
                lEndPoint++
            }
        }


    }

    /***
     * This helper function resolves chapters from the specified document and returns a list of chapters.
     * Parent - parseResponseToChapters();
     *
     * @param parsedDocument
     * @return
     */
    private fun resolveChaptersFromParsedDocument(parsedDocument: Document, request: RequestWrapper): List<DbChapter>
    {
        val lChapterList = ArrayList<DbChapter>()
        val lChapterElements = parsedDocument.select("ul.chapter-list").select("li")
        var lNumChapters = lChapterElements.size

        var lChapterUrl: String
        var lChapterTitle: String
        var lChapterDate: String

        for (iChapterElement in lChapterElements)
        {
            lChapterUrl = iChapterElement.select("a").attr("href")
            lChapterTitle = iChapterElement.select("span").first().text()
            lChapterDate = iChapterElement.select("span")[1].text()

            lChapterList.add(DbChapter(lChapterUrl, request.manga.title, lChapterTitle, lChapterDate, lNumChapters, request.manga
                    .link, SourceKey))

            lNumChapters--
        }

        return lChapterList
    }

    companion object
    {
        val TAG = FunManga::class.java.simpleName
        val URL = "https://www.funmanga.com"
        val SourceKey = "FunManga"
    }
}



